### DISCUSSION TITLE
Oqtane App State Management

### DISCUSSION THREAD

**Discussion by Sharafudeen** [oqtane/oqtane.framework#2097] [Q&A] [answered]
Hi,

How Oqtane manage App State features like in Fluxor / Flux / Redux library which can use in Blazor?

**Reference:** https://youtu.be/Vn6dKN_hTrs

**Comment by sbwalker** [comment, marked as answer]
@Sharafudeen you don't need any of those third party solutions to manage state in Blazor - there are are plenty of native solutions available including cascading parameters, scoped services, etc... Oqtane utilizes these to maintain state and makes them available to developers who are building extensions ie. SiteState, PageState, ModuleState. I would encourage you to build an Oqtane extension as you will find most of the common Blazor development challenges outlined in generic online articles and videos are handled for you.

**Reply by curlyfro** [reply]
i may also have a need for Fluxor in my Oqtane app.  now that I've got OpenId working, I'm starting to think about converting one of my Blazor clients over to Oqtane.  in my existing Blazor app I have a use case where i change user roles and i have Fluxor event listeners change the UI (in Oqtane, the UI would be across several modules) based on user roles.  @sbwalker are you saying there's i can do this with a PageState extension?

**Comment by sbwalker** [comment]
@curlyfro I would need to understand what you are trying to achieve in your application. It sounds like you need some sort of notifications when events occur... but your description it is not clear in terms of the location of the logical workloads. If you consider the following diagram there are 3 distinct types of interactions - client-client, client-server, and server-server. Which of these are applicable to your application?  

![image](https://user-images.githubusercontent.com/4840590/161428195-dc644873-cc03-4dca-85f3-cde4769d414b.png)

**Reply by curlyfro** [reply]
our apps have the following roles: super admin, admin, editor, author, contributor, and subscriber.  our independent modules are role-based -- meaning they all have the UI switching logic to display data based on the role.  as a super admin and partial admin, you can change roles to see what the UI view looks like for that role.  role changes can affect root menu item visibility and the data/UI of each module.  role changes will trigger API requests to get data for that role.  so in your diagram, I guess my interactions are client-client and client-server.

**Reply by sbwalker** [reply]
I am not sure if I fully understand but it sounds like you created a solution which allows a user to "masquerade" as another user so that they can see what the UI looks like when logged in as that user. I would personally recommend against that approach as I have seen too many security vulnerabilities introduced by such a "feature". It is much better to simply create multiple users - each with the appropriate role assignments - and actually login as the specific user to get their UX experience. Yes, it might make testing a bit more cumbersome however it is much safer from a security perspective. But that is just my opinion.

In regards to client-to-client interaction you might be interested in https://www.oqtane.org/blog/!/40/inotifypropertychanged

**Reply by curlyfro** [reply]
unfortunately, this masquerading users feature is not my decision :(

i read the blog post.  i like this pattern better.
